diff --git a/dlls/winepulse.drv/Makefile.in b/dlls/winepulse.drv/Makefile.in
index 84fb540..068eed0 100644
--- a/dlls/winepulse.drv/Makefile.in
+++ b/dlls/winepulse.drv/Makefile.in
@@ -1,9 +1,10 @@
 MODULE    = winepulse.drv
-IMPORTS   = dxguid uuid winmm user32 advapi32 ole32
-EXTRALIBS = $(PULSE_LIBS) $(PTHREAD_LIBS)
-EXTRAINCL = $(PULSE_CFLAGS)
+UNIXLIB   = winepulse.so
+IMPORTS   = dxguid uuid winmm user32 advapi32 ole32 version
+UNIX_LIBS    = $(PULSE_LIBS) $(PTHREAD_LIBS) -lm
+UNIX_CFLAGS  = $(PULSE_CFLAGS)
 
 EXTRADLLFLAGS = -mcygwin
 
 C_SRCS = \
-	mmdevdrv.c
+	mmdevdrv.c 

--- a/include/wine/test.h
+++ b/include/wine/test.h
@@ -28,6 +28,13 @@
 #include <winbase.h>
 #include <wine/debug.h>
 
+#ifdef __WINE_CONFIG_H
+#error config.h should not be used in Wine tests
+#endif
+#ifdef __WINE_WINE_UNICODE_H
+#error wine/unicode.h should not be used in Wine tests
+#endif
+
 #ifndef INVALID_FILE_ATTRIBUTES
 #define INVALID_FILE_ATTRIBUTES  (~0u)
 #endif
--- /dev/null
+++ b/include/wine/unicode.h
@@ -0,0 +1,178 @@
+/*
+ * Wine internal Unicode definitions
+ *
+ * Copyright 2000 Alexandre Julliard
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#if 0
+#pragma makedep install
+#endif
+
+#ifndef __WINE_WINE_UNICODE_H
+#define __WINE_WINE_UNICODE_H
+
+#include <stdarg.h>
+
+#include <windef.h>
+#include <winbase.h>
+#include <winnls.h>
+#include <winternl.h>
+
+#ifdef __WINE_USE_MSVCRT
+#error This file should not be used with msvcrt headers
+#endif
+
+#ifndef WINE_UNICODE_INLINE
+#define WINE_UNICODE_INLINE static FORCEINLINE
+#endif
+
+WINE_UNICODE_INLINE WCHAR tolowerW( WCHAR ch )
+{
+    return RtlDowncaseUnicodeChar( ch );
+}
+
+WINE_UNICODE_INLINE WCHAR toupperW( WCHAR ch )
+{
+    return RtlUpcaseUnicodeChar( ch );
+}
+
+WINE_UNICODE_INLINE int isspaceW( WCHAR wc )
+{
+    unsigned short type;
+    GetStringTypeW( CT_CTYPE1, &wc, 1, &type );
+    return type & C1_SPACE;
+}
+
+WINE_UNICODE_INLINE unsigned int strlenW( const WCHAR *str )
+{
+    const WCHAR *s = str;
+    while (*s) s++;
+    return s - str;
+}
+
+WINE_UNICODE_INLINE WCHAR *strcpyW( WCHAR *dst, const WCHAR *src )
+{
+    WCHAR *p = dst;
+    while ((*p++ = *src++));
+    return dst;
+}
+
+WINE_UNICODE_INLINE WCHAR *strcatW( WCHAR *dst, const WCHAR *src )
+{
+    strcpyW( dst + strlenW(dst), src );
+    return dst;
+}
+
+WINE_UNICODE_INLINE WCHAR *strrchrW( const WCHAR *str, WCHAR ch )
+{
+    WCHAR *ret = NULL;
+    do { if (*str == ch) ret = (WCHAR *)(ULONG_PTR)str; } while (*str++);
+    return ret;
+}
+
+WINE_UNICODE_INLINE int strcmpiW( const WCHAR *str1, const WCHAR *str2 )
+{
+    for (;;)
+    {
+        int ret = tolowerW(*str1) - tolowerW(*str2);
+        if (ret || !*str1) return ret;
+        str1++;
+        str2++;
+    }
+}
+
+WINE_UNICODE_INLINE int strncmpiW( const WCHAR *str1, const WCHAR *str2, int n )
+{
+    int ret = 0;
+    for ( ; n > 0; n--, str1++, str2++)
+        if ((ret = tolowerW(*str1) - tolowerW(*str2)) || !*str1) break;
+    return ret;
+}
+
+WINE_UNICODE_INLINE LONG strtolW( LPCWSTR s, LPWSTR *end, INT base )
+{
+    BOOL negative = FALSE, empty = TRUE;
+    LONG ret = 0;
+
+    if (base < 0 || base == 1 || base > 36) return 0;
+    if (end) *end = (WCHAR *)s;
+    while (isspaceW(*s)) s++;
+
+    if (*s == '-')
+    {
+        negative = TRUE;
+        s++;
+    }
+    else if (*s == '+') s++;
+
+    if ((base == 0 || base == 16) && s[0] == '0' && (s[1] == 'x' || s[1] == 'X'))
+    {
+        base = 16;
+        s += 2;
+    }
+    if (base == 0) base = s[0] != '0' ? 10 : 8;
+
+    while (*s)
+    {
+        int v;
+
+        if ('0' <= *s && *s <= '9') v = *s - '0';
+        else if ('A' <= *s && *s <= 'Z') v = *s - 'A' + 10;
+        else if ('a' <= *s && *s <= 'z') v = *s - 'a' + 10;
+        else break;
+        if (v >= base) break;
+        if (negative) v = -v;
+        s++;
+        empty = FALSE;
+
+        if (!negative && (ret > MAXLONG / base || ret * base > MAXLONG - v))
+            ret = MAXLONG;
+        else if (negative && (ret < (LONG)MINLONG / base || ret * base < (LONG)(MINLONG - v)))
+            ret = MINLONG;
+        else
+            ret = ret * base + v;
+    }
+
+    if (end && !empty) *end = (WCHAR *)s;
+    return ret;
+}
+
+NTSYSAPI int __cdecl _vsnwprintf(WCHAR*,size_t,const WCHAR*,__ms_va_list);
+
+static inline int WINAPIV snprintfW( WCHAR *str, size_t len, const WCHAR *format, ...)
+{
+    int retval;
+    __ms_va_list valist;
+    __ms_va_start(valist, format);
+    retval = _vsnwprintf(str, len, format, valist);
+    __ms_va_end(valist);
+    return retval;
+}
+
+static inline int WINAPIV sprintfW( WCHAR *str, const WCHAR *format, ...)
+{
+    int retval;
+    __ms_va_list valist;
+    __ms_va_start(valist, format);
+    retval = _vsnwprintf(str, MAXLONG, format, valist);
+    __ms_va_end(valist);
+    return retval;
+}
+
+#undef WINE_UNICODE_INLINE
+
+#endif  /* __WINE_WINE_UNICODE_H */
--- a/include/Makefile.in
+++ b/include/Makefile.in
@@ -850,6 +850,7 @@
 	wine/strmbase.h \
 	wine/svcctl.idl \
 	wine/test.h \
+	wine/unicode.h \
 	wine/unixlib.h \
 	wine/vulkan.h \
 	wine/vulkan_driver.h \


diff --git a/libs/uuid/uuid.c b/libs/uuid/uuid.c
index ce97520..1ba0397 100644
--- a/libs/uuid/uuid.c
+++ b/libs/uuid/uuid.c
@@ -92,7 +92,6 @@ DEFINE_GUID(GUID_NULL,0,0,0,0,0,0,0,0,0,0,0);
 #include "comsvcs.h"
 #include "directmanipulation.h"
 #include "devguid.h"
-#include "propkey.h"
 
 #define __WINE_INIT_NPI_MODULEID
 #include "netiodef.h"

diff --git a/tools/winegcc/winegcc.c b/tools/winegcc/winegcc.c
index a850cce..5f2098d 100644
--- a/tools/winegcc/winegcc.c
+++ b/tools/winegcc/winegcc.c
@@ -1554,6 +1554,8 @@ int main(int argc, char **argv)
     memset(&opts, 0, sizeof(opts));
     opts.target = init_argv0_target( argv[0] );
     opts.pic = 1;
+    
+    if (is_pe_target( &opts )) opts.use_msvcrt = 1;
 
     /* determine the processor type */
     if (strendswith(argv[0], "winecpp")) opts.processor = proc_cpp;
@@ -1984,8 +1986,6 @@ int main(int argc, char **argv)
     if (opts.processor == proc_cpp) linking = 0;
     if (linking == -1) error("Static linking is not supported\n");
 
-    if (is_pe_target( &opts )) opts.use_msvcrt = 1;
-
     if (opts.files.count == 0 && !opts.fake_module) forward(&opts);
     else if (linking) build(&opts);
     else compile(&opts, lang);
     

diff --git a/tools/makedep.c b/tools/makedep.c
index 532145d..625c7f1 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -27,6 +27,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <stdarg.h>
+#include <signal.h>
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
@@ -197,6 +198,7 @@ struct makefile
     const char     *staticlib;
     const char     *importlib;
     const char     *unixlib;
+    int             use_msvcrt;
     int             data_only;
     int             is_win16;
     int             is_exe;
@@ -596,17 +598,6 @@ static int is_multiarch( unsigned int arch )
 }
 
 
-/*******************************************************************
- *         is_using_msvcrt
- *
- * Check if the files of a makefile use msvcrt by default.
- */
-static int is_using_msvcrt( struct makefile *make )
-{
-    return make->module || make->testdll;
-}
-
-
 /*******************************************************************
  *         arch_module_name
  */
@@ -865,7 +856,7 @@ static struct incl_file *add_generated_source( struct makefile *make, const char
     file->basename = xstrdup( filename ? filename : name );
     file->filename = obj_dir_path( make, file->basename );
     file->file->flags = FLAG_GENERATED;
-    file->use_msvcrt = is_using_msvcrt( make );
+    file->use_msvcrt = make->use_msvcrt;
     list_add_tail( &make->sources, &file->entry );
     if (make == include_makefile)
     {
@@ -1614,7 +1605,7 @@ static struct incl_file *add_src_file( struct makefile *make, const char *name )
 
     memset( file, 0, sizeof(*file) );
     file->name = xstrdup(name);
-    file->use_msvcrt = is_using_msvcrt( make );
+    file->use_msvcrt = make->use_msvcrt;
     file->is_external = !!make->extlib;
     list_add_tail( &make->sources, &file->entry );
     if (make == include_makefile)
@@ -1812,12 +1803,13 @@ static void add_generated_sources( struct makefile *make )
     unsigned int i, arch;
     struct incl_file *source, *next, *file, *dlldata = NULL;
     struct strarray objs = get_expanded_make_var_array( make, "EXTRA_OBJS" );
+    int multiarch = archs.count > 1 && make->use_msvcrt;
 
     LIST_FOR_EACH_ENTRY_SAFE( source, next, &make->sources, struct incl_file, entry )
     {
         for (arch = 0; arch < archs.count; arch++)
         {
-            if (!is_multiarch( arch )) continue;
+            if (!arch != !multiarch) continue;
             if (source->file->flags & FLAG_IDL_CLIENT)
             {
                 file = add_generated_source( make, replace_extension( source->name, ".idl", "_c.c" ), NULL, arch );
@@ -1921,7 +1913,7 @@ static void add_generated_sources( struct makefile *make )
     {
         for (arch = 0; arch < archs.count; arch++)
         {
-            if (!is_multiarch( arch )) continue;
+            if (!arch != !multiarch) continue;
             file = add_generated_source( make, "testlist.o", "testlist.c", arch );
             add_dependency( file->file, "wine/test.h", INCL_NORMAL );
             add_all_includes( make, file, file->file );
@@ -2175,6 +2167,7 @@ static int is_crt_module( const char *file )
  */
 static const char *get_default_crt( const struct makefile *make )
 {
+    if (!make->use_msvcrt) return NULL;
     if (make->module && is_crt_module( make->module )) return NULL;  /* don't add crt import to crt dlls */
     return !make->testdll && (!make->staticlib || make->extlib) ? "ucrtbase" : "msvcrt";
 }
@@ -2364,7 +2357,6 @@ static struct strarray get_source_defines( struct makefile *make, struct incl_fi
         strarray_add( &ret, strmake( "-I%s", root_src_dir_path( "include/msvcrt" )));
         for (i = 0; i < make->include_paths.count; i++)
             strarray_add( &ret, strmake( "-I%s", make->include_paths.str[i] ));
-        strarray_add( &ret, get_crt_define( make ));
     }
     strarray_addall( &ret, make->define_args );
     strarray_addall( &ret, get_expanded_file_local_var( make, obj, "EXTRADEFS" ));
@@ -2425,9 +2417,7 @@ static void output_winegcc_command( struct makefile *make, unsigned int arch )
         output_filename( tools_path( make, "winebuild" ));
     }
     output_filenames( target_flags[arch] );
-    if (arch) return;
-    output_filename( "-mno-cygwin" );
-    output_filenames( lddll_flags );
+    if (!arch) output_filenames( lddll_flags );
 }
 
 
@@ -2744,8 +2734,7 @@ static void output_source_rc( struct makefile *make, struct incl_file *source, c
     if (source->file->flags & FLAG_RC_HEADER) return;
     if (source->file->flags & FLAG_GENERATED) strarray_add( &make->clean_files, source->name );
     if (linguas.count && (source->file->flags & FLAG_RC_PO)) po_dir = "po";
-    for (arch = 0; arch < archs.count; arch++)
-        if (!make->disabled[arch]) strarray_add( &make->res_files[arch], res_file );
+    for (arch = 0; arch < archs.count; arch++) strarray_add( &make->res_files[arch], res_file );
     if (source->file->flags & FLAG_RC_PO)
     {
         strarray_add( &make->pot_files, strmake( "%s.pot", obj ));
@@ -2782,8 +2771,7 @@ static void output_source_mc( struct makefile *make, struct incl_file *source, c
     char *obj_path = obj_dir_path( make, obj );
     char *res_file = strmake( "%s.res", obj );
 
-    for (arch = 0; arch < archs.count; arch++)
-        if (!make->disabled[arch]) strarray_add( &make->res_files[arch], res_file );
+    for (arch = 0; arch < archs.count; arch++) strarray_add( &make->res_files[arch], res_file );
     strarray_add( &make->pot_files, strmake( "%s.pot", obj ));
     output( "%s.pot %s.res: %s", obj_path, obj_path, source->filename );
     output_filename( tools_path( make, "wmc" ));
@@ -2808,7 +2796,6 @@ static void output_source_mc( struct makefile *make, struct incl_file *source, c
  */
 static void output_source_res( struct makefile *make, struct incl_file *source, const char *obj )
 {
-    if (make->disabled[source->arch]) return;
     strarray_add( &make->res_files[source->arch], source->name );
 }
 
@@ -2820,10 +2807,10 @@ static void output_source_idl( struct makefile *make, struct incl_file *source,
 {
     struct strarray defines = get_source_defines( make, source, obj );
     struct strarray headers = empty_strarray;
-    struct strarray deps = empty_strarray;
     struct strarray multiarch_targets[MAX_ARCHS] = { empty_strarray };
     const char *dest;
     unsigned int i, arch;
+    int multiarch;
 
     if (find_include_file( make, strmake( "%s.h", obj ))) source->file->flags |= FLAG_IDL_HEADER;
     if (!source->file->flags) return;
@@ -2847,10 +2834,10 @@ static void output_source_idl( struct makefile *make, struct incl_file *source,
     for (i = 0; i < ARRAY_SIZE(idl_outputs); i++)
     {
         if (!(source->file->flags & idl_outputs[i].flag)) continue;
+        multiarch = (make->use_msvcrt && archs.count > 1);
         for (arch = 0; arch < archs.count; arch++)
         {
-            if (!is_multiarch( arch )) continue;
-            if (make->disabled[arch]) continue;
+            if (!arch != !multiarch) continue;
             dest = strmake( "%s%s%s", arch_dirs[arch], obj, idl_outputs[i].ext );
             if (!find_src_file( make, dest )) strarray_add( &make->clean_files, dest );
             strarray_add( &multiarch_targets[arch], dest );
@@ -2859,12 +2846,9 @@ static void output_source_idl( struct makefile *make, struct incl_file *source,
 
     for (arch = 0; arch < archs.count; arch++)
     {
-        struct strarray arch_deps = empty_strarray;
-
-        if (!arch) strarray_addall( &arch_deps, headers );
-        strarray_addall( &arch_deps, multiarch_targets[arch] );
-        if (!arch_deps.count) continue;
-        output_filenames_obj_dir( make, arch_deps );
+        if (multiarch_targets[arch].count + (arch ? 0 : headers.count) == 0) continue;
+        if (!arch) output_filenames_obj_dir( make, headers );
+        output_filenames_obj_dir( make, multiarch_targets[arch] );
         output( ":\n" );
         output( "\t%s%s -o $@", cmd_prefix( "WIDL" ), tools_path( make, "widl" ) );
         output_filenames( target_flags[arch] );
@@ -2875,18 +2859,15 @@ static void output_source_idl( struct makefile *make, struct incl_file *source,
         output_filenames( get_expanded_file_local_var( make, obj, "EXTRAIDLFLAGS" ));
         output_filename( source->filename );
         output( "\n" );
-        strarray_addall( &deps, arch_deps );
     }
 
-    if (deps.count)
-    {
-        output_filenames_obj_dir( make, deps );
-        output( ":" );
-        output_filename( tools_path( make, "widl" ));
-        output_filename( source->filename );
-        output_filenames( source->dependencies );
-        output( "\n" );
-    }
+    output_filenames_obj_dir( make, headers );
+    for (arch = 0; arch < archs.count; arch++) output_filenames_obj_dir( make, multiarch_targets[arch] );
+    output( ":" );
+    output_filename( tools_path( make, "widl" ));
+    output_filename( source->filename );
+    output_filenames( source->dependencies );
+    output( "\n" );
 
     if (source->importlibdeps.count)
     {
@@ -3135,18 +3116,16 @@ static void output_source_one_arch( struct makefile *make, struct incl_file *sou
 {
     const char *obj_name;
 
-    if (make->disabled[arch] && !(source->file->flags & FLAG_C_IMPLIB)) return;
-
     if (arch)
     {
         if (source->file->flags & FLAG_C_UNIX) return;
-        if (!is_using_msvcrt( make ) && !make->staticlib && !(source->file->flags & FLAG_C_IMPLIB)) return;
+        if (!make->use_msvcrt && !make->staticlib && !(source->file->flags & FLAG_C_IMPLIB)) return;
     }
     else if (source->file->flags & FLAG_C_UNIX)
     {
         if (!*dll_ext) return;
     }
-    else if (archs.count > 1 && is_using_msvcrt( make ) &&
+    else if (archs.count > 1 && make->use_msvcrt &&
              !(source->file->flags & FLAG_C_IMPLIB) &&
              (!make->staticlib || make->extlib)) return;
 
@@ -3276,8 +3255,6 @@ static void output_fake_module( struct makefile *make )
     unsigned int arch = 0;  /* fake modules are always native */
     const char *spec_file = NULL, *name = strmake( "%s%s", arch_pe_dirs[arch], make->module );
 
-    if (make->disabled[arch]) return;
-
     if (!make->is_exe) spec_file = src_dir_path( make, replace_extension( make->module, ".dll", ".spec" ));
 
     strarray_add( &make->all_targets[arch], name );
@@ -3316,8 +3293,6 @@ static void output_module( struct makefile *make, unsigned int arch )
     char *spec_file = NULL;
     unsigned int i;
 
-    if (make->disabled[arch]) return;
-
     if (!make->is_exe) spec_file = src_dir_path( make, replace_extension( make->module, ".dll", ".spec" ));
 
     if (!make->data_only)
@@ -3331,6 +3306,12 @@ static void output_module( struct makefile *make, unsigned int arch )
         strarray_addall( &all_libs, add_import_libs( make, &dep_libs, default_imports, IMPORT_TYPE_DEFAULT, arch ) );
         if (!arch) strarray_addall( &all_libs, libs );
 
+        if (!make->use_msvcrt)
+        {
+            strarray_addall( &all_libs, get_expanded_make_var_array( make, "UNIX_LIBS" ));
+            strarray_addall( &all_libs, libs );
+        }
+
         if (delay_load_flags[arch])
         {
             for (i = 0; i < make->delayimports.count; i++)
@@ -3423,8 +3404,6 @@ static void output_unix_lib( struct makefile *make )
     struct strarray unix_libs = add_unix_libraries( make, &unix_deps );
     unsigned int arch = 0;  /* unix libs are always native */
 
-    if (make->disabled[arch]) return;
-
     strarray_add( &make->all_targets[arch], make->unixlib );
     add_install_rule( make, make->module, arch, make->unixlib,
                       strmake( "p%s%s", arch_install_dirs[arch], make->unixlib ));
@@ -3562,7 +3541,7 @@ static void output_test_module( struct makefile *make, unsigned int arch )
     output( ": %s", obj_dir_path( make, testmodule ));
     if (parent)
     {
-        char *parent_module = arch_module_name( make->testdll, arch );
+        char *parent_module = arch_module_name( make->testdll, parent->use_msvcrt ? arch : 0 );
         output_filename( obj_dir_path( parent, parent_module ));
         if (parent->unixlib) output_filename( obj_dir_path( parent, parent->unixlib ));
     }
@@ -3813,7 +3792,12 @@ static void output_sources( struct makefile *make )
     }
     else if (make->module)
     {
-        for (arch = 0; arch < archs.count; arch++) if (is_multiarch( arch )) output_module( make, arch );
+        if (!make->use_msvcrt) output_module( make, 0 );
+        else
+        {
+            for (arch = 0; arch < archs.count; arch++)
+                if (is_multiarch( arch )) output_module( make, arch );
+        }
         if (make->unixlib) output_unix_lib( make );
         if (make->importlib) for (arch = 0; arch < archs.count; arch++) output_import_lib( make, arch );
         if (make->is_exe && !make->is_win16 && *dll_ext && strendswith( make->module, ".exe" ))
@@ -4059,6 +4043,16 @@ static void output_stub_makefile( struct makefile *make )
     const char *make_var = strarray_get_value( &top_makefile->vars, "MAKE" );
     unsigned int i, arch;
 
+    if (make->obj_dir) create_dir( make->obj_dir );
+
+    output_file_name = obj_dir_path( make, "Makefile" );
+    output_file = create_temp_file( output_file_name );
+
+    output( "# Auto-generated stub makefile; all rules forward to the top-level makefile\n\n" );
+
+    if (make_var) output( "MAKE = %s\n\n", make_var );
+    output( "all:\n" );
+
     for (arch = 0; arch < archs.count; arch++)
         if (make->all_targets[arch].count) strarray_add_uniq( &targets, "all" );
 
@@ -4076,16 +4070,6 @@ static void output_stub_makefile( struct makefile *make )
         strarray_add( &targets, "testclean" );
     }
 
-    if (!targets.count && !make->clean_files.count) return;
-
-    output_file_name = obj_dir_path( make, "Makefile" );
-    output_file = create_temp_file( output_file_name );
-
-    output( "# Auto-generated stub makefile; all rules forward to the top-level makefile\n\n" );
-
-    if (make_var) output( "MAKE = %s\n\n", make_var );
-
-    output( "all:\n" );
     output_filenames( targets );
     output_filenames( make->clean_files );
     output( ":\n" );
@@ -4255,9 +4239,13 @@ static void load_sources( struct makefile *make )
     }
     make->is_win16   = strarray_exists( &make->extradllflags, "-m16" );
     make->data_only  = strarray_exists( &make->extradllflags, "-Wb,--data-only" );
+    make->use_msvcrt = (make->module || make->testdll || make->is_win16) &&
+                       !strarray_exists( &make->extradllflags, "-mcygwin" );
     make->is_exe     = strarray_exists( &make->extradllflags, "-mconsole" ) ||
                        strarray_exists( &make->extradllflags, "-mwindows" );
 
+    if (make->use_msvcrt) strarray_add_uniq( &make->extradllflags, "-mno-cygwin" );
+
     if (make->module)
     {
         /* add default install rules if nothing was specified */
@@ -4314,6 +4302,8 @@ static void load_sources( struct makefile *make )
 
     add_generated_sources( make );
 
+    if (make->use_msvcrt) strarray_add( &make->define_args, get_crt_define( make ));
+
     LIST_FOR_EACH_ENTRY( file, &make->includes, struct incl_file, entry ) parse_file( make, file, 0 );
     LIST_FOR_EACH_ENTRY( file, &make->sources, struct incl_file, entry ) get_dependencies( file, file );
 
@@ -4413,7 +4403,11 @@ int main( int argc, char *argv[] )
     if (argc > 1) fatal_error( "Directory arguments not supported in this mode\n" );
 
     atexit( cleanup_files );
-    init_signals( exit_on_signal );
+    signal( SIGTERM, exit_on_signal );
+    signal( SIGINT, exit_on_signal );
+#ifdef SIGHUP
+    signal( SIGHUP, exit_on_signal );
+#endif
 
     for (i = 0; i < HASH_SIZE; i++) list_init( &files[i] );
     for (i = 0; i < HASH_SIZE; i++) list_init( &global_includes[i] );
